File: core/src/main/kotlin/com/auto_switch_ime/core/ime/input/ImeProviderRegistry.kt
package com.auto_switch_ime.core.ime.input

import com.auto_switch_ime.core.ImeConfig
import com.auto_switch_ime.core.ImeException
import com.auto_switch_ime.core.ImeProvider
import com.auto_switch_ime.core.ImeType
import com.auto_switch_ime.core.util.Logger

class ImeProviderRegistry {
    private val providers = mutableMapOf<ImeType, (ImeConfig, Logger) -> ImeProvider>()

    fun register(type: ImeType, factory: (ImeConfig, Logger) -> ImeProvider) {
        providers[type] = factory
    }

    fun createProvider(config: ImeConfig, logger: Logger): ImeProvider {
        val factory = providers[config.type]
            ?: throw ImeException.ProviderNotFound(config.type)
        return factory(config, logger)
    }

    fun supportedTypes(): Set<ImeType> = providers.keys.toSet()
}

/**
 * 平台入口注册已实现的输入法级 Provider，Coordinator 只依赖 ImeGateway。
 */
object ImeProviderFactory {
    private val registry = ImeProviderRegistry()

    fun register(type: ImeType, factory: (ImeConfig, Logger) -> ImeProvider) {
        registry.register(type, factory)
    }

    fun createProvider(config: ImeConfig, logger: Logger): ImeProvider {
        return registry.createProvider(config, logger)
    }

    fun supportedTypes(): Set<ImeType> = registry.supportedTypes()
}

File: vscode/src/ime/input/ImeProviderRegistry.ts
import { ImeConfig, ImeProvider, ImeType } from '../../core/types';

type ProviderFactory = (config: ImeConfig) => ImeProvider;

export class ImeProviderRegistry {
  private readonly providers = new Map<ImeType, ProviderFactory>();

  register(type: ImeType, factory: ProviderFactory): void {
    this.providers.set(type, factory);
  }

  create(config: ImeConfig): ImeProvider {
    const factory = this.providers.get(config.type);
    if (!factory) {
      throw new Error(`IME provider not found for type: ${config.type}`);
    }
    return factory(config);
  }

  supportedTypes(): ImeType[] {
    return [...this.providers.keys()];
  }
}
